sklearn/utils/validation.py

```diff
@@ -239,19 +240,36 @@ def _ensure_sparse_format(spmatrix, accept_sparse, dtype, copy,
     spmatrix_converted : scipy sparse matrix.
         Matrix that is ensured to have an allowed type.
     """
-    if accept_sparse in [None, False]:
-        raise TypeError('A sparse matrix was passed, but dense '
-                        'data is required. Use X.toarray() to '
-                        'convert to a dense numpy array.')
     if dtype is None:
         dtype = spmatrix.dtype
 
     changed_format = False
-    if (isinstance(accept_sparse, (list, tuple))
-            and spmatrix.format not in accept_sparse):
-        # create new with correct sparse
-        spmatrix = spmatrix.asformat(accept_sparse[0])
-        changed_format = True
+
+    if isinstance(accept_sparse, six.string_types):
+        accept_sparse = [accept_sparse]
+
+    if accept_sparse in [None, False]:
+        if accept_sparse is None:
+            warnings.warn(
+                    "Passing None to parameter 'accept_sparse' is "
+                    "deprecated in 0.19. Use False instead.",
```

Comment:
We should deprecate that in ``check_array`` imho, as the deprecation warning should be raised even if a dense array is passed.